sdk/middleware/sqecho/echo.go

```diff
@@ -0,0 +1,73 @@
+package sqecho
+
+import (
+	"context"
+
+	"github.com/labstack/echo"
+	"github.com/sqreen/go-agent/sdk"
+)
+
+// Middleware is Sqreen's middleware function for Echo to monitor and protect
+// the requests Echo receives. It creates and stores the HTTP request record
+// both into the Echo and request contexts so that it can be later accessed from
+// handlers using `sdk.FromContext()` to perform SDK calls.
+//
+// Note that Echo's context does not implement the `context.Context` interface,
+// so `sdk.FromContext()` cannot be used with it, hence `FromContext()` in this
+// package to access the SDK context value from Echo's context.
+// `sdk.FromContext()` can still be used on the request context.
+//
+//	e := echo.New()
+//	e.Use(sqecho.Middleware())
+//
+//	e.GET("/", func(c echo.Context) {
+//		// Accessing the SDK through Echo's context
+//		sqecho.FromContext(c).TrackEvent("my.event.one")
+//		foo(c.Request())
+//	}
+//
+//	func foo(req *http.Request) {
+//		// Accessing the SDK through the request context
+//		sdk.FromContext(req.Context()).TrackEvent("my.event.two")
+//	}
+//
+func Middleware() echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			// Create a new request record for this request.
+			req := c.Request()
+			sqreen := sdk.NewHTTPRequestRecord(req)
+			defer sqreen.Close()
```

Comment:
Called before actually returning from the function, with or without a panic (~go exceptions).
So the actual flow here is:

1. call `next(c)`.
1. call the deferred functions 
1. return

But this is really just about being called even in case of a panic.